text/0000-release-process.md

```diff
@@ -0,0 +1,194 @@
+                        |  development  | ---> | code freeze | ---> | release |
+                        +---------------+      +-------------+      +---------+
+```
+
+### Terminology
+
+#### User Acceptance Testing
+
+User acceptance testing (UAT) is the process of manually verifying that new features and fixes are **complete** and 
+**don't introduce issues**. Typically, there are two strategies:
+
+1. Rolling UAT - User acceptance is done per issue as the issue is completed.
+2. Feature-Complete UAT - User acceptance is done when all features are complete.
+
+More commonly Feature-Complete User Acceptance is used in practice because any additional changes could have unexpected
+side-effects invalidating prior completed UATs. In our case, Feature-Complete UAT fit perfectly during the
+[code freeze](#code-freeze) phase.
+
+#### Code Freeze
+
+A code freeze for a particular release is a period of time when changes are analyzed and vetted before being added to
+a release.
+
+The following take place:
+
+1. At the beginning, a release branch is created (`release/<version>`).
+2. At the beginning, a [change control board](#change-control-board) is organized on a daily basis to review additional
+requested changes.
+3. During, only changes approved by the CCB are merged into the release branch.
+
+#### Change Control Board
+
+A change control board would consist of core developers that, on a daily basis, organize changes that might
+be suitable for [release branches](#release-branches) during the code freeze.
+
+The board members typically decide on whether a change should be included on the release based on the following criteria:
+
+- Impact
+- Effort
+- Risk
+
+#### Release branches
+
+Release branches are protected branches that will be created at the beginning of a code freeze. These branches are
+protected, similar to `master`, if not more so depending on the [change control board](#change-control-board) members. 
+
+The release branch will be merged into `master` upon release.
+
+```text
+                 master
+
+                   o
+                   |
+                   |    release/1.2.3
+                   |
+                   |          o commit: fix/critical-or-minor
+                   |          |
+                   |          |
+commit: feature/Y  o          |
+                   |          |
+                   |          |
+                   |          |
+                   |          |
+                   +----------+ <--+ start code freeze
+                   |
+                   |
+                   |
+                   |
+commit: feature/X  o
+                   |
+                   |
+                   +
+```
+
+#### Migration Guide
+
+A migration guide is a document which details breaking changes for migrating from prior versions.
+
+### Release Execution
+
+Upon execution of the release the following will take place:
+
+- A github release will be created containing the following:
```

Comment:
What about formalizing the cutting of an `-rc.1` release when we have a code freeze. This lets some of our consumers test out candidate changes. We could cut them as pre-releases and if the CCB lets in more changes, we could cut `-rc.2`, etc?

Reply:
That's a great idea! I hadn't thought about the possibility of providing pre-release builds. I've got a few initial thoughts:

- I wonder if it's something we want to take on right now or defer onto a separate RFC. 
- My initial concern would be that it might not be as stable as we'd like with such a short release cycle. 
- Given that we would have a short release cycle I'm not sure how much value pre-release builds would provide. We would need very tight feedback loops for reporting issues and addressing them.

Reply:
I think we could have a followup RFC on the process of release candidates, for sure. And yes I agree that sort turnaround makes this not super useful.